sumpy/e2e.py

```diff
@@ -424,21 +517,120 @@ def __call__(self, queue, **kwargs):
         :arg m2l_precomputed_exprs:
         :arg m2l_translation_vectors:
         """
-        knl = self.get_cached_optimized_kernel()
-
         m2l_translation_vectors = kwargs.pop("m2l_translation_vectors")
         # "1" may be passed for rscale, which won't have its type
         # meaningfully inferred. Make the type of rscale explicit.
         src_rscale = m2l_translation_vectors.dtype.type(kwargs.pop("src_rscale"))
 
+        m2l_precomputed_exprs = kwargs.pop("m2l_precomputed_exprs")
+        result_dtype = m2l_precomputed_exprs.dtype
+
+        knl = self.get_cached_optimized_kernel(result_dtype=result_dtype)
+
         return knl(queue,
                 src_rscale=src_rscale,
                 m2l_translation_vectors=m2l_translation_vectors,
+                m2l_precomputed_exprs=m2l_precomputed_exprs,
                 **kwargs)
 
 # }}}
 
 
+# {{{ class for calculating the FFT of multipole expansion for M2L
+
+class E2EFromCSRWithFFTPreprocess(E2EFromCSR):
+    """Implements preprocessing the multipole expansions
+    for M2L with FFT.
```

Comment:
```suggestion
    """Computes the preprocessed multipoled coefficients for accelerated
    M2L.
```

Reply:
Fixed in 8045c01db3af4f63f7aa4bb19ed95998f79e304b